core/containerExec.go

```diff
@@ -0,0 +1,357 @@
+			var sharingMode llb.CacheMountSharingMode
+			switch mnt.CacheSharingMode {
+			case CacheSharingModeShared:
+				sharingMode = llb.CacheMountShared
+			case CacheSharingModePrivate:
+				sharingMode = llb.CacheMountPrivate
+			case CacheSharingModeLocked:
+				sharingMode = llb.CacheMountLocked
+			default:
+				return nil, errors.Errorf("invalid cache mount sharing mode %q", mnt.CacheSharingMode)
+			}
+
+			mountOpts = append(mountOpts, llb.AsPersistentCacheDir(mnt.CacheVolumeID, sharingMode))
+		}
+
+		if mnt.Tmpfs {
+			mountOpts = append(mountOpts, llb.Tmpfs())
+		}
+
+		if mnt.Readonly {
+			mountOpts = append(mountOpts, llb.Readonly)
+		}
+
+		runOpts = append(runOpts, llb.AddMount(mnt.Target, srcSt, mountOpts...))
+	}
+
+	if opts.InsecureRootCapabilities {
+		runOpts = append(runOpts, llb.Security(llb.SecurityModeInsecure))
+	}
+
+	fsSt, err := container.FSState()
+	if err != nil {
+		return nil, fmt.Errorf("fs state: %w", err)
+	}
+
+	execMDOpt, err := execMD.AsConstraintsOpt()
+	if err != nil {
+		return nil, fmt.Errorf("execution metadata: %w", err)
+	}
+	runOpts = append(runOpts, execMDOpt)
+	execSt := fsSt.Run(runOpts...)
+
+	marshalOpts := []llb.ConstraintsOpt{
+		llb.Platform(platform.Spec()),
+		execMDOpt,
+	}
+	execDef, err := execSt.Root().Marshal(ctx, marshalOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("marshal root: %w", err)
+	}
+
+	container.FS = execDef.ToPB()
+
+	metaDef, err := execSt.GetMount(buildkit.MetaMountDestPath).Marshal(ctx, marshalOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("get meta mount: %w", err)
+	}
+
+	container.Meta = metaDef.ToPB()
+
+	for i, mnt := range mounts {
+		if mnt.Tmpfs || mnt.CacheVolumeID != "" {
+			continue
+		}
+
+		mountSt := execSt.GetMount(mnt.Target)
+
+		// propagate any changes to regular mounts to subsequent containers
+		execMountDef, err := mountSt.Marshal(ctx, marshalOpts...)
+		if err != nil {
+			return nil, fmt.Errorf("propagate %s: %w", mnt.Target, err)
+		}
+
+		mounts[i].Source = execMountDef.ToPB()
+	}
+
+	container.Mounts = mounts
+
+	// set image ref to empty string
+	container.ImageRef = ""
+
+	return container, nil
+}
+
+func (container *Container) MetaFileContents(ctx context.Context, filePath string) (string, error) {
+	if container.Meta == nil {
+		ctr, err := container.WithExec(ctx, ContainerExecOpts{})
+		if err != nil {
+			return "", err
+		}
+		return ctr.MetaFileContents(ctx, filePath)
+	}
+
+	file := NewFile(
+		container.Query,
+		container.Meta,
+		path.Join(buildkit.MetaMountDestPath, filePath),
+		container.Platform,
+		container.Services,
+	)
+
+	content, err := file.Contents(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	return string(content), nil
+}
+
+func metaMount(stdin string) (llb.State, string) {
+	// because the shim might run as non-root, we need to make a world-writable
+	// directory first and then make it the base of the /dagger mount point.
+	//
+	// TODO(vito): have the shim exec as the other user instead?
```

Comment:
Ah good catch, actually this should be entirely invalidated now since the metamount isn't even actually mounted into the container anymore.

We still need to "mount" it from the perspective of LLB in order for its cached result to be tied to the result of the actual exec, but we remove it from the OCI spec in the executor and just write to it directly from there. Kind of nicer in general since I suppose previously it was probably possible for user execs to try shenanigans w/ write access to it.

So I'll just remove this comment now.